pkg/cmd/infra.go

```diff
@@ -0,0 +1,373 @@
+			}
+			switch args[0] {
+			case "orphan":
+				switch args[1] {
+				case "list":
+					var rs []string
+					pathTerraformState := downloadTerraformFiles("infra", targetReader) + "/terraform.tfstate"
+					buf, err := ioutil.ReadFile(pathTerraformState)
+					if err != nil || len(buf) < 64 {
+						fmt.Println("Could not read terraform.tfstate: " + pathTerraformState)
+						os.Exit(2)
+					}
+					terraformstate := string(buf)
+
+					shoot, err := FetchShootFromTarget(target)
+					checkError(err)
+					infraType := shoot.Spec.Provider.Type
+
+					switch infraType {
+						case "aws":
+							rs = getAWSInfraResources()
+						case "gcp":
+						case "azure":
+							rs = getAzureInfraResources()
+						case "alicloud":
+						case "openstack":
+						default:
+							return errors.New("infra type not found")
+					}
+
+					getOrphanInfraResources(rs, terraformstate)
+					fmt.Printf("\nsearched %s\n", pathTerraformState)
+				default:
+					fmt.Println("command must be in the format: infra [orphan] [list]")
+				}
+			default:
+				fmt.Println("command must be in the format: infra [orphan] [list]")
+			}
+
+			return nil
+		},
+		ValidArgs: []string{"orphan"},
+	}
+}
+
+// getOrphanInfraResources list orphan infra resources on targeted cluster
+func getOrphanInfraResources(rs []string, terraformstate string) error {
+	var hasOrphan bool
+
+	if (len(rs) < 1) {
+		fmt.Println("No infra resources available")
+		os.Exit(2)
+	}
+
+	fmt.Printf("(%d) infra resources found: \n%s\n", len(rs), rs)
+	for _, rsid := range rs {
+		if !strings.Contains(terraformstate, rsid) {
+			fmt.Printf("\nOrphan: resource id %s not found in terraform state\n", rsid)
+			hasOrphan = true
+		}
+	}
+
+	if (!hasOrphan) {
+		fmt.Printf("\nNo orphan resource found\n")
+	}
+
+	return nil
+}
+
+func getAWSInfraResources() []string {
+	var target Target
+	ReadTarget(pathTarget, &target)
+	rs := make([]string, 0, 16)
+	shoottag := "shoot--" + target.Target[1].Name + "--" + target.Target[2].Name
+
+	// fetch shoot vpc resources
+	arguments := "aws ec2 describe-vpcs --filter Name=tag:kubernetes.io/cluster/" + shoottag + ",Values=1"
+	captured := capture()
+	operate("aws", arguments)
+	capturedOutput, err := captured()
+	checkError(err)
+	re, _ := regexp.Compile(`VPCS.*(vpc-[a-z0-9]*)`)
+	values := re.FindAllStringSubmatch(capturedOutput, -1)
+	if len(values) > 0 {
+		for i:=0; i < len(values); i++ {
+			rs = append(rs, values[i][1])
+		}
+	}
+	// fetch shoot subnet resources
+	arguments = "aws ec2 describe-subnets --filter Name=tag:kubernetes.io/cluster/" + shoottag + ",Values=1"
+	captured = capture()
+	operate("aws", arguments)
+	capturedOutput, err = captured()
+	checkError(err)
+	re, _ = regexp.Compile(`SUBNETS.*:subnet\/(subnet-[a-z0-9]*)`)
+	values = re.FindAllStringSubmatch(capturedOutput, -1)
+	if len(values) > 0 {
+		for i:=0; i < len(values); i++ {
+			rs = append(rs, values[i][1])
+		}
+	}
```

Comment:
There is a lot of duplicate code in this function. Most of the commands below executed share the same pattern with one or more regexes to apply.  It would be nice to extract it to some helper function, if possible